Source/WebCore/page/Navigation.cpp

```diff
@@ -122,14 +122,17 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
     RefPtr currentItem = frame()->history().currentItem();
     if (!currentItem)
         return;
-    if (previousWindow) {
+    // For main frames we can still rely on the page b/f list. However for subframes we need below logic to not lose the bookkeeping done in the previous window.
+    if (previousWindow && !frame()->isMainFrame()) {
         Ref previousNavigation = previousWindow->protectedNavigation();
-        if (previousNavigation->m_entries.size() && navigationType && navigationType == NavigationNavigationType::Reload && frame()->document()->protectedSecurityOrigin()->isSameOriginDomain(previousWindow->document()->protectedSecurityOrigin())) {
+        if (previousNavigation->m_entries.size() && (navigationType == NavigationNavigationType::Reload || navigationType == NavigationNavigationType::Push) && frame()->document()->protectedSecurityOrigin()->isSameOriginAs(previousWindow->document()->protectedSecurityOrigin())) {
```

Comment:
This looks a bit messy and hard to follow, especially since there is a check for NavigationNavigationType::Reload inside.  Instead, please use a switch statement or make a function that has a name that indicates what it is about ::Reload and ::Push that meets the condition here, or otherwise make the logic easier to follow.

Also, is the switch from isSameOriginDomain to isSameOriginAs intentional?

Reply:
I believe SameOrigin is correct:

> newSHEs will have originally come from [getting session history entries for the navigation API](https://html.spec.whatwg.org/multipage/browsing-the-web.html#getting-session-history-entries-for-the-navigation-api), and thus each newSHE will be contiguous [same](https://html.spec.whatwg.org/multipage/browsers.html#same-origin) [origin](https://html.spec.whatwg.org/multipage/browsing-the-web.html#document-state-origin) with initialSHE.